q2_annotate/_utils.py

```diff
@@ -96,13 +99,49 @@ def _calculate_md5_from_file(file_path: str) -> str:
     return md5_hash.hexdigest()
 
 
-def get_feature_lengths(features: MAGSequencesDirFmt) -> pd.DataFrame:
+def get_feature_lengths(
+        features: Union[
+            MAGSequencesDirFmt,
+            MultiMAGSequencesDirFmt,
+            ContigSequencesDirFmt,
+            DNASequencesDirectoryFormat
+        ]
+) -> pd.DataFrame:
     """Calculate lengths of features in a feature data object."""
     ids, lengths = [], []
-    for _id, fp in features.feature_dict().items():
-        sequences = skbio.io.read(fp, format='fasta', verify=False)
-        ids.append(_id)
-        lengths.append(sum(len(seq) for seq in sequences))
+    per_sequence = False
+
+    # for FeatureData[Sequence]
+    if isinstance(features, DNASequencesDirectoryFormat):
+        sample_dict = {"": {"": os.path.join(str(features), 'dna-sequences.fasta')}}
+        per_sequence = True
+
+        # For SampleData[Contigs]
+    elif isinstance(features, ContigSequencesDirFmt):
+        file_dict = features.sample_dict()
+        sample_dict = {"": file_dict}  # Create fake sample for sample_dict
+        per_sequence = True
+
+    # For SampleData[MAGs]
+    elif isinstance(features, MultiMAGSequencesDirFmt):
+        sample_dict = features.sample_dict()
+
+    # For FeatureData[MAG]
+    else:
+        file_dict = features.feature_dict()
+        sample_dict = {"": file_dict}  # Create fake sample for sample_dict
+
+    for sample_id, file_dict in sample_dict.items():
```

Comment:
Since sample_id is not used, let's do:
```suggestion
    for _, file_dict in sample_dict.items():
```